### PR TITLE
GAWB-2623: Endpoint to GET a list of clusters that are not deleted

### PR DESCRIPTION
**Pull request by ansingh7115**
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/resources/swagger/api-docs.yaml [hunk 1]
```diff
@@ -55,6 +55,44 @@ paths:
...
+        '500':
+          description: Internal Error
+          schema:
+            $ref: '#/definitions/ErrorReport'
+      security:
+        - googleoauth:
+            - openid
+            - email
+            - profile
+  '/api/clusters/includeDeleted':
```
Comment (ansingh7115) [inline src/main/resources/swagger/api-docs.yaml]: is this alright or would we prefer `/api/clusters/includeDeleted`

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 2]
```diff
@@ -87,10 +87,10 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig, gdDAO: Datap
     }
   }
 
-  def listClusters(labelMap: Map[String, String]): Future[Seq[Cluster]] = {
+  def listClusters(labelMap: Map[String, String], includeDeleted: Boolean = false): Future[Seq[Cluster]] = {
```
Comment (ansingh7115) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Instead, I could also just have 2 functions `listClusters(labelMap)` that gives all the clusters, and `listExistingClusters(labelMap)` that gives all clusters that aren't deleted.

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala [hunk 3]
```diff
@@ -270,9 +272,36 @@ class LeonardoServiceSpec extends TestKit(ActorSystem("leonardotest")) with Flat
...
+    leo.listClusters(Map.empty, true).futureValue.toSet shouldBe Set(cluster1, cluster2)
+    leo.listClusters(Map.empty, false).futureValue.toSet shouldBe Set(cluster1, cluster2)
+
+    val clusterName3 = s"cluster-${UUID.randomUUID.toString}"
+    val cluster3 = leo.createCluster(googleProject, clusterName3, testClusterRequest.copy(labels = Map("a" -> "b", "foo" -> "bar"))).futureValue
+    leo.deleteCluster(googleProject, clusterName3)
+
+    db.inTransaction(dataAccess =>
+      dataAccess.clusterQuery.updateClusterStatus(cluster3.googleId, ClusterStatus.Deleted)
+    )
```
Comment (ansingh7115) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala]: I'm marking the third cluster as deleted directly in the db - not sure if there's a better way?

**Comment by ansingh7115** [comment]
I haven't written any api tests for this new endpoint because we would need a way to mark clusters as deleted directly through the service (rather than marking them for deletion). Should I write a function that directly deletes a cluster expressly for testing and plop it in LeoService? Or is there a better way?

File: src/main/resources/swagger/api-docs.yaml (hunk 1 above)
Comment (helgridly) [reply src/main/resources/swagger/api-docs.yaml]: I'd prefer `&includeDeleted`. I'm okay if we want to tell Swagger the parameter is mandatory as long as Leo behaves correctly (i.e. in all four cases: no query string, `&includeDeleted`, `&includeDeleted=true`, and `&includeDeleted=false`).

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala [hunk 4]
```diff
@@ -32,10 +32,12 @@ object ClusterStatus extends Enumeration {
   val Unknown, Creating, Running, Updating, Error, Deleting, Deleted = Value
   val activeStatuses = Set(Unknown, Creating, Running, Updating)
   val monitoredStatuses = Set(Unknown, Creating, Updating, Deleting)
+  val existingStatuses = Set(Unknown, Creating, Running, Updating, Error, Deleting)
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: I see what you're going for here, but I think this needs a better name. `aliveStatuses`? Open to suggestions...

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 2 above)
Comment (helgridly) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: This is fine.

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala (hunk 3 above)
Comment (helgridly) [reply src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala]: Given that there isn't a LeoService method to do it, this is fine.

File: src/main/resources/swagger/api-docs.yaml (hunk 1 above)
Comment (rtitle) [reply src/main/resources/swagger/api-docs.yaml]: +1, I agree with @helgridly

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala (hunk 4 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeonardoModel.scala]: `visibleStatuses`?

File: src/main/resources/swagger/api-docs.yaml [hunk 5]
```diff
@@ -71,6 +71,12 @@ paths:
             labels as top-level query string parameters. For instance: GET /api/clusters?key1=val1&key2=val2.
           required: false
           type: string
+        - in: query
+          name: includeDeleted
+          description: ugh
```
Comment (jmthibault79) [inline src/main/resources/swagger/api-docs.yaml]: heh

Comment (rtitle) [reply src/main/resources/swagger/api-docs.yaml]: hey that's my reaction too when working with swagger!

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala [hunk 6]
```diff
@@ -273,6 +275,33 @@ class LeonardoServiceSpec extends TestKit(ActorSystem("leonardotest")) with Flat
...
+    val clusterName2 = s"cluster-${UUID.randomUUID.toString}"
+    val cluster2 = leo.createCluster(googleProject, clusterName2, testClusterRequest.copy(labels = Map("a" -> "b", "foo" -> "bar"))).futureValue
+
+    leo.listClusters(Map.empty).futureValue.toSet shouldBe Set(cluster1, cluster2)
+    leo.listClusters(Map.empty).futureValue.toSet shouldBe Set(cluster1, cluster2)
+
+    val clusterName3 = s"cluster-${UUID.randomUUID.toString}"
+    val cluster3 = leo.createCluster(googleProject, clusterName3, testClusterRequest.copy(labels = Map("a" -> "b", "foo" -> "bar"))).futureValue
+
+    db.inTransaction(dataAccess =>
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala]: you can use `dbFutureValue` here - and that way you don't need to keep the `db` around either

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala [hunk 7]
```diff
@@ -115,9 +115,15 @@ trait ClusterComponent extends LeoComponent {
       }
     }
 
-    def listByLabels(labelMap: Map[String, String]): DBIO[Seq[Cluster]] = {
-      val query = if (labelMap.isEmpty) {
+    def listByLabels(labelMap: Map[String, String], includeDeleted: Boolean): DBIO[Seq[Cluster]] = {
+      val clusterStatusQuery = if (includeDeleted)
         clusterQueryWithLabels
+      else
+        clusterQueryWithLabels.filter { _._1.status inSetBind ClusterStatus.existingStatuses.map(_.toString) }
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala]: What about `filterNot { status === 'Deleted' }` instead?  Do we really need `existingStatuses`?

**Comment by jmthibault79** [comment]
👍

**Pull request merged**